src/main/java/com/epam/esm/service/TagService.java

```diff
@@ -17,43 +21,29 @@
 @Service
 public class TagService {
 
-    private final GiftCertificateTagRepository giftCertificateTagRepository;
-
+    private final TagRepository tagRepository;
 
     @Autowired
-    public TagService(GiftCertificateTagRepository giftCertificateTagRepository) {
-        this.giftCertificateTagRepository = giftCertificateTagRepository;
+    public TagService(TagRepository tagRepository) {
+        this.tagRepository = tagRepository;
     }
 
-    /**
-     *
-     * @param tagName name of the tag to be saved.
-     * @return
-     * if tag name is not valid, returns bad request
-     * if tag already exists, returns bad request
-     * if tag does not exist, but cannot be saved, return bad request
-     * if it is saved, return CREATED and tag saved
-     */
+
     public ResponseEntity<?> saveTag(String tagName) {
         if (tagName == null || tagName.isEmpty()) {
             return ResponseEntity.badRequest().body(new ErrorDTO("Tag name is required", TAG_BAD_REQUEST));
         }
-        Tag tag = giftCertificateTagRepository.getTagByName(tagName);
 
-        if (tag != null) {
-
-            String message = TAG_ALREADY_EXISTS.formatted(tag.getId());
+        Optional<Tag> possibleTag = tagRepository.findByName(tagName);
+        if (possibleTag.isPresent()) {
+            String message = TAG_ALREADY_EXISTS.formatted(possibleTag.get().getId());
             return ResponseEntity.badRequest().body(new ErrorDTO(message, TAG_BAD_REQUEST));
         }
 
-        Tag tagSaved = giftCertificateTagRepository.saveTag(tagName);
-        if (tagSaved == null) {
-            return ResponseEntity.status(BAD_REQUEST).body(new ErrorDTO(TAG_COULD_NOT_BE_SAVED, TAG_BAD_REQUEST));
-        }
-
-        Tag tagResponse = giftCertificateTagRepository.getTagById(tagSaved.getId());
-
-        return ResponseEntity.status(CREATED).body(tagResponse);
+        Tag tag = new Tag();
+        tag.setName(tagName);
+        Tag savedTag = tagRepository.save(tag);
```

Comment:
You would need to process exceptions here: https://stackoverflow.com/questions/71300043/spring-data-jpa-how-to-implement-proper-exception-handling - as example.
Basically, if your database has problems JPA repository would throw some exception at you. You need to catch it and convert it into proper response, so that user would not receive raw exception in Postman.

Reply:
Resolved